### PR TITLE
SKU metric

### PR DESCRIPTION
**Pull request by jackdelahunt**
# Description
Added a new metric for the currrent SKU configuration for RHOAM. The metric stores the current config, and if the config is changing what it is changing to. This metric also has an annotation in Grafana so when the SKU config is changed it can be seen in Grafana.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [ ] This change requires a documentation update 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added a test case that will be used to verify my changes 
- [ ] Verified independently on a cluster by reviewer

## Verify
- Log into a cluster with the changes (ask me for details)
- Change the SKU configuration of RHOAM -> [here](https://github.com/jackdelahunt/integreatly-operator/blob/170b7aaed0c4573b02c386365766038ae602af71/Makefile#L363)
- Verify that the annotation appears on Grafana by checking a dashboard

image

### CODE REVIEW

File: controllers/rhmi/rhmi_controller.go [hunk 1]
```diff
@@ -342,6 +342,8 @@ func (r *RHMIReconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
 	// Check for stage complete to avoid setting the metric when installation is happening
 	if string(installation.Status.Stage) == "complete" {
 		metrics.SetRhmiVersions(string(installation.Status.Stage), installation.Status.Version, installation.Status.ToVersion, installation.CreationTimestamp.Unix())
+
+		metrics.SetActiveSKU(string(installation.Status.Stage), installation.Status.SKU, installation.Status.ToSKU)
```
Comment (laurafitzgerald) [inline controllers/rhmi/rhmi_controller.go:346]:
```suggestion
		metrics.SetSKUs(string(installation.Status.Stage), installation.Status.SKU, installation.Status.ToSKU)
```

File: controllers/rhmi/rhmi_controller.go [hunk 2]
```diff
@@ -410,10 +412,12 @@ func (r *RHMIReconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
 		if installation.Spec.RebalancePods {
 			r.reconcilePodDistribution(installation)
 		}
+    
 		if installation.Spec.Type == string(rhmiv1alpha1.InstallationTypeManagedApi) {
 			if installationSKU.IsUpdated() {
 				installation.Status.SKU = installationSKU.GetName()
 				installation.Status.ToSKU = ""
+        metrics.SetActiveSKU(string(installation.Status.Stage), installation.Status.SKU, installation.Status.ToSKU)
```
Comment (laurafitzgerald) [inline controllers/rhmi/rhmi_controller.go:420]:
```suggestion
        metrics.SetSKUs(string(installation.Status.Stage), installation.Status.SKU, installation.Status.ToSKU)
```

File: main.go [hunk 3]
```diff
@@ -58,6 +58,7 @@ func init() {
 	customMetrics.Registry.MustRegister(integreatlymetrics.RHOAMVersion)
 	customMetrics.Registry.MustRegister(integreatlymetrics.RHOAMStatus)
 	customMetrics.Registry.MustRegister(integreatlymetrics.ThreeScaleUserAction)
+	customMetrics.Registry.MustRegister(integreatlymetrics.ActiveSKU)
```
Comment (laurafitzgerald) [inline main.go:61]:
```suggestion
	customMetrics.Registry.MustRegister(integreatlymetrics.SKU)
```
OR 
```suggestion
	customMetrics.Registry.MustRegister(integreatlymetrics.RHOAMSKU)
```

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -101,6 +101,18 @@ var (
 			"action",
 		},
 	)
+
+	ActiveSKU = prometheus.NewGaugeVec(
```
Comment (laurafitzgerald) [inline pkg/metrics/metrics.go:105]:
```suggestion
	SKU = prometheus.NewGaugeVec(
```

File: pkg/products/monitoring/dashboards/clusterResources.go [hunk 5]
```diff
@@ -29,6 +29,22 @@ func GetMonitoringGrafanaDBClusterResourcesJSON(installationName string) string
...
 				"useValueForTime": false
+			}, 
+			{
+				"datasource": "Prometheus",
+				"enable": true,
+				"expr": "count by (stage,sku,toSKU)(active_sku{toSKU!=\"\"})",
+				"hide": false,
+				"iconColor": "#FADE2A",
+				"limit": 100,
+				"name": "Active SKU",
```
Comment (laurafitzgerald) [inline pkg/products/monitoring/dashboards/clusterResources.go:40]:
```suggestion
				"name": "SKU Change (million per day)",
```

File: pkg/products/monitoring/dashboards/clusterResources.go [hunk 6]
```diff
@@ -29,6 +29,22 @@ func GetMonitoringGrafanaDBClusterResourcesJSON(installationName string) string
...
+				"showIn": 0,
+				"step": "",
+				"tagKeys": "stage,sku,toSKU",
+				"tags": "",
+				"titleFormat": "Active SKU",
```
Comment (laurafitzgerald) [inline pkg/products/monitoring/dashboards/clusterResources.go:45]:
```suggestion
				"titleFormat": "SKU Change (million per day)",
```

**Review by laurafitzgerald — COMMENTED** [review]
Hi @jackdelahunt I haven't verified, just looked at the code. Overall the change looks good. I left some naming change suggestions. The main one there being not using `Active SKU` as the metric will refer to both active sku and current sku. Like the version the metric also exposes a toversion and version so we can follow the same naming pattern also.

**Comment by laurafitzgerald** [comment]
@jackdelahunt running `make code/fix` will resolve the failing test.

**Comment by jackdelahunt** [comment]
New commit makes the changes you proposed

**Comment by laurafitzgerald** [comment]
thanks for the change @jackdelahunt would you mind sharing it to the team channel for verification if someone is free. I've still a bit to do with a pr this afternoon.

**Comment by laurafitzgerald** [comment]
Hi @jackdelahunt you'll have to drop that last commit before you ask someone to verify. If it's needed for someone to verify it to run it's better to write steps on what value to replace where otherwise once the verification is done you will need to drop the commit from the pr and this will kick off a new set of tests.

**Pull request closed without merging**